include/circt/Dialect/Moore/MooreOps.td

```diff
@@ -1104,17 +1117,27 @@ def UnionExtractOp : MooreOp<"union_extract"> {
     ```
     See IEEE 1800-2017 § 7.3.1 "Packed unions"
   }];
-  let arguments = (ins StrAttr:$memberName, UnpackedType:$input);
+  let arguments = (ins StrAttr:$fieldName, UnpackedType:$input);
   let results = (outs UnpackedType:$result);
   let assemblyFormat = [{
-  $input `,`   $memberName  attr-dict `:`
+  $input `,`   $fieldName  attr-dict `:`
+  type($input) `->` type($result)
+  }];
+}
+
+def UnionExtractRefOp : MooreOp<"union_extract_ref"> {
+  let summary = "Union Extract operation";
+  let arguments = (ins StrAttr:$fieldName, RefType:$input);
+  let results = (outs RefType:$result);
+  let assemblyFormat = [{
+  $input `,`   $fieldName  attr-dict `:`
   type($input) `->` type($result)
   }];
 }
```

Comment:
Same as for the struct case: it would be great to have `$input` be a `AnyUnionType` and `AnyUnionRefType`, and to have `TypesMatchWith<...>` constraints that check if the `$result` type is the same as the accessed variant/field of the `$input` union. (Wrapped in `ref<...>` for the `*_ref` op.)

Reply:
As mentioned, I wrote a verifier to fix this.